crates/ide/src/hover.rs

```diff
@@ -111,100 +113,179 @@ pub(crate) fn hover(
-                        Either::Right(it) => Definition::Macro(it),
-                    })
-                // attributes, require special machinery as they are mere ident tokens
-                } else if let Some(attr) = token.ancestors().find_map(ast::Attr::cast) {
-                    // lints
-                    if let res@Some(_) = try_hover_for_lint(&attr, &token) {
-                        return res;
-                    // derives
+            match node {
+                ast::Name(name) => NameClass::classify(sema, &name).map(|class| match class {
+                    NameClass::Definition(it) | NameClass::ConstReference(it) => it,
+                    NameClass::PatFieldShorthand { local_def, field_ref: _ } => Definition::Local(local_def),
+                }),
+                ast::NameRef(name_ref) => NameRefClass::classify(sema, &name_ref).map(|class| match class {
+                    NameRefClass::Definition(def) => def,
+                    NameRefClass::FieldShorthand { local_ref: _, field_ref } => {
+                        Definition::Field(field_ref)
+                    }
+                }),
+                ast::Lifetime(lifetime) => NameClass::classify_lifetime(&sema, &lifetime).map_or_else(
+                    || {
+                        NameRefClass::classify_lifetime(&sema, &lifetime).and_then(|class| match class {
+                            NameRefClass::Definition(it) => Some(it),
+                            _ => None,
+                        })
+                    },
+                    NameClass::defined,
+                ),
+                _ => {
+                    // intra-doc links
+                    // FIXME: move comment + attribute special cases somewhere else to simplify control flow,
+                    // hopefully simplifying the return type of this function in the process
+                    // (the `Break`/`Continue` distinction is needed to decide whether to use fallback hovers)
+                    //
+                    // FIXME: hovering the intra doc link to `Foo` not working:
+                    //
+                    // #[identity]
+                    // trait Foo {
+                    //    /// [`Foo`]
+                    // fn foo() {}
+                    if token.kind() == COMMENT {
+                        add_to_seen_definitions = false;
+                        cov_mark::hit!(no_highlight_on_comment_hover);
+    let (attributes, def) = doc_attributes(sema, node)?;
+    let (docs, doc_mapping) = attributes.docs_with_rangemap(sema.db)?;
+    let (idl_range, link, ns) =
+        extract_definitions_from_docs(&docs).into_iter().find_map(|(range, link, ns)| {
+            let mapped = doc_mapping.map(range)?;
+            (mapped.file_id == file_id.into() && mapped.value.contains(offset)).then(||(mapped.value, link, ns))
+        })?;
+    range_override = Some(idl_range);
+    Some(match resolve_doc_path_for_def(sema.db,def, &link,ns)? {
```

Comment:
indentation missing, rustfmt doesn't work here since this is inside a macro invocation